### PR TITLE
Comments: adds the comment date to the collapsed view

### PR DESCRIPTION
**Pull request by rodrigoi**
This PR adds the comment date to the collapsed view.

I'm presenting two options for consideration. The standard time formatted to the user's timezone:

screen shot 2017-10-06 at 14 38 53

And the human readable time period of time we use in places like the reader.

screen shot 2017-10-06 at 14 40 33

There's a small duplication of code, and I had to add a couple of props to the already populated prop list of `comment-detail-header.jsx`, but it'll be fixed on #18406 so I'm not concerned.

References:

p3fqKv-5LC-p2

### CODE REVIEW

**Comment by megsfulton** [comment]
👍

**Comment by kwight** [comment]
I noticed the human-readable version only goes back 24hrs presumably; I get `20h ago` on one comment, with the next being the date (Sept 29) – shouldn't I be able to see `x days ago`?

**Comment by rodrigoi** [comment]
@kwight weird. It should go up to "days ago":

https://github.com/Automattic/wp-calypso/blob/2082817a62c6e52f314e2f183bb247de68d84f2e/client/lib/human-date/index.js#L40

Can you add a screenshot?

**Comment by kwight** [comment]
screen shot 2017-10-06 at 1 05 55 pm

**Comment by kwight** [comment]
Oh! Maybe it only goes up to 7 days back; changing the date of the Facebook comment to Oct 2nd gives `4d ago`. Sept 28 would be 8 days ago.

File: client/blocks/comment-detail/comment-detail-header.jsx [hunk 1]
```diff
@@ -116,6 +130,9 @@ export const CommentDetailHeader = ( {
 									} ) }
 								</Emojify>
 							</div>
+							<div className="comment-detail__author-info-element is-timestamp">
```
Comment (Copons) [inline client/blocks/comment-detail/comment-detail-header.jsx]: Not sure if `is-timestamp` is the correct naming here, as I wouldn't think of it as a modifier.
Maybe something like `comment-detail__author-info-timestamp`, while a total mouthful, could be more apt?

**Comment by Copons** [comment]
@kwight Ah yup, it goes 7 days back at most: https://github.com/Automattic/wp-calypso/blob/master/client/lib/human-date/index.js#L44

I've tried both approaches and I like the relative dates better, even though I find two details slightly off putting:
- 7 days feels a bit too strict as a cut off date. Maybe 2 weeks or even 1 month would be better?
- In most cases we'll lose the time, which can be useful for blogs with many comments (aka power users 😄 ).

File: client/blocks/comment-detail/comment-detail-header.jsx (hunk 1 above)
Comment (rodrigoi) [reply client/blocks/comment-detail/comment-detail-header.jsx]: probably. I hate **BEM**. I was so, SO, tempted to just do `style={{ fontSize: '12px' }}`...

Comment (rodrigoi) [reply client/blocks/comment-detail/comment-detail-header.jsx]: Updated on cf9a01bc861ee58e6d04fd00a479b193f18f9b81.

**Comment by Copons** [comment]
Thanks for the change!

The code LGTM, but I'd like to have some additional 👀 at how the relative dates look like before merging.
Do you think it might make sense to enable it by default in this PR so that people can test in the live branch?

**Comment by rodrigoi** [comment]
@Copons, yeah, considering that's the style is being scrutinized the most. Updated on 2dff9280be247bcbdfbde858abaee1e5e5de436a. Just refresh the live branch.

**Comment by megsfulton** [comment]
Couple of things: 

1) Could we update the relative time stamp to use "days", "hours" and "mins" instead of d, h and m? 
2) Can we increase the vertical space between the post title and date to be equal with the space between the post title and author. I adjusted the line height in the inspector to 1.6 and that seemed about right.   

screen shot 2017-10-09 at 10 37 38 am


3) This might belong in its own PR but anywhere we're using the color #537994 we should update it to #4f748e so that we're using colors that are consistent with the [design guidelines](https://wordpress.com/design-handbook/colors/)

**Comment by Copons** [comment]
> Could we update the relative time stamp to use "days", "hours" and "mins" instead of d, h and m?

I guess we could expand the `humanDate` lib with an options argument that defaults to the current display mode to avoid messing with other sections already using it.

**Comment by roundhill** [comment]
Looking good 👍 

Can we update the height of the placeholder rows as well? It doesn't quite look right now that we have larger rows:

oct-09-2017 16-57-44

To replicate that just change pages in the navigation at the bottom.

**Comment by rodrigoi** [comment]
well, after some research I've found that `humanDate` is not the only way Calypso manages relative time periods. Updating the code to use `i18n-calypso` instance of `moment` we can go back... FOREVER. And use correct language, with full words.

screen shot 2017-10-09 at 22 55 15

Also updated the preview styles and increased the vertical space between lines.

@megsfulton I'll prepare a new PR with color changes. Maybe it'll require more than one PR, because we'll require some changes on the base components line `Button`.

**Comment by Copons** [comment]
@roundhill @rodrigoi Leave the placeholder to me, after we merge this.
I've been meaning to revamp its code a bit, because it's been a pain since the beginning.

---

@megsfulton @rodrigoi Wait a bit with the color!
We're using `$gray-text-min` (`#537994`) which is the minimum contrast color on white background.
```sass
$gray-text-min: darken( $gray, 18% ); //#537994
```

In [_colors.scss](https://github.com/Automattic/wp-calypso/blob/master/assets/stylesheets/shared/_colors.scss#L18) there is also a mention of `#4f748e `:
```sass
darken( $gray, 20% ) //#4f748e
```

It is slightly darker than `$gray-text-min` so there shouldn't be a11y concerns, but I wonder which is the most correct to use.
`$gray-text-min` is used 126 times across Calypso, while `darken( $gray, 20% )` 70.
I guess these results aren't conclusive enough, besides the fact that `$gray-text-min` has its own variable and `darken( ... )` doesn't.

IIRC the last text color update was @drw158 work, so we're out of luck right now. Maybe @folletto knows more about this?

**Comment by folletto** [comment]
Yep I was involved in that decision, so I can answer :) 

The difference is semantics: if it's text, using `$gray-text-min` is better as future tweaks to _text_ will be easier. :)

FWIW "is used 126 times across Calypso" is not useful in this case unfortunately as we didn't do a replacement pass for all uses when the decision was done, but we're slowly introducing it. So even in situations where it's more appropriate, it might still be using the other, less semantic, variable.

tl;dr: if it's text, prefer `$gray-text-min`; if it's any other UI element, do not use at all, ever, `$gray-text-min`. :)

(Aside: unfortunately this change hasn't been propagate in the Design Handbook yet as it's harder to update).

**Comment by Copons** [comment]
> tl;dr: if it's text, prefer `$gray-text-min`;

So in this case `$gray-text-min` is correct. 👍 

> if it's any other UI element, do not use at all, ever, `$gray-text-min`. :)

😱 

_[run away checking every component I've ever styled]_

**Comment by roundhill** [comment]
Tried out the new changes, I think this is looking great 👍

**Comment by kwight** [comment]
Are we intentionally using the human-friendly version on the list, and the actual on the comment detail view?

screen shot 2017-10-10 at 10 42 23 am

If we're using the human-friendly version, users should still have _some_ way of seeing the exact time (like @Copons mentioned for power users with many comments per hour). Two ways of handling this I've seen are hovers with the exact time, or tapping/clicking to alternate between formats (better for mobile, but involves tapping on little text?). We could certainly leave it as-is for this PR, and explore in a later PR.

**Comment by megsfulton** [comment]
@folletto thanks for clarifying on the colors! I'll be sure to use the right one in my designs. 

>Are we intentionally using the human-friendly version on the list, and the actual on the comment detail view?

Yes? I think this behavior is okay.

One thing that might help though is using the exact date for anything over a month old (or pick a time period shorter than that). There's not much benefit to seeing "a month ago", "a year ago" over and over.

**Comment by rodrigoi** [comment]
@megsfulton what do you think it's a good cutoff for having relative times vs regular dates? I'm leaning towards having relative times for everything newer than a month.
We can also add the regular date as a tooltip. We should check if Calypso has a special semantic to indicate the presence of a tooltip.

**Comment by Copons** [comment]
> tapping/clicking to alternate between formats

This is super cool, but wouldn't work in this case, with the whole card header being clickable to toggle expand. 🙁

Anyway, when I mentioned the power users, I forgot about the full date in the expanded view. 😄 
With the actual possibility to double check the precise time of a comment already implemented, I've got no more issues against the relative dates—except the cutoff decision, which is not an issue per se though.

**Comment by folletto** [comment]
> One thing that might help though is using the exact date for anything over a month old (or pick a time period shorter than that). There's not much benefit to seeing "a month ago", "a year ago" over and over.

+100


> when I mentioned the power users, I forgot about the full date in the expanded view. 

Yeah, I was to say "ok to tap to show the tooltip for human readable dates", but since it's in the expanded view already, I think there's no need other than the cut-off after 29 days :)

**Comment by Copons** [comment]
About the third line in the comment excerpt, it should be a super easy change, maybe worth adding to this PR.

A couple of changes to [these rules](https://github.com/Automattic/wp-calypso/blob/master/client/blocks/comment-detail/style.scss#L194-L238) are enough:

```sass
// from `min-height: 38px;`
min-height: 57px;

// from `-webkit-line-clamp: 2;`
-webkit-line-clamp: 3;
```

Just, we might want to keep 2 lines on `<960px`.

File: client/blocks/comment-detail/comment-detail-header.jsx [hunk 2]
```diff
@@ -36,6 +39,8 @@ export const CommentDetailHeader = ( {
 	isEditMode,
 	isExpanded,
 	postTitle,
+	site,
```
Comment (Copons) [inline client/blocks/comment-detail/comment-detail-header.jsx]: We don't seem to be passing `site` through from `CommentDetail`.

Comment (rodrigoi) [reply client/blocks/comment-detail/comment-detail-header.jsx]: great catch! Thanks. Fixed on 51167355aec01d71122f16037d210cb6b1438dba.

**Comment by megsfulton** [comment]
Agree with the cutoff being somewhere around the month mark for relative dates. No strong opinion on whether that's 28, 29 or 30 days.

**Comment by rodrigoi** [comment]
@Copons changes made, per our discussion over Slack 😉  on a500f6073dd58baa866799b7c0ff31d310009b52.

@megsfulton I've added a cutoff period of a relative _month_. It will try to go back one calendar month relative to the current date, or round to the closest day.

**Comment by kwight** [comment]
This all looks great! The only detail that odds me out now is seeing the relative and precise dates together; we go from general relative to very specific precision. Can we drop the time from the listing, and just keep it for the detail part? Anyone else feel the same?

**Comment by megsfulton** [comment]
>Can we drop the time from the listing, and just keep it for the detail part? Anyone else feel the same?

We're intentionally adding it in because without it the newest/oldest sort isn't particularly informative.

I don't find it that strange to have the mixed timestamps in a list. A lot mobile inbox / mail apps tend to do relative dates falling back to date stamps and it's never struck me as odd. However, one could maybe argue for getting rid of the time stamp at the list level to make things a little cleaner / simpler.

**Comment by kwight** [comment]
> However, one could maybe argue for getting rid of the time stamp at the list level to make things a little cleaner / simpler.

Oh, yes, this is all I meant :)

**Comment by megsfulton** [comment]
>Oh, yes, this is all I meant :)

Oh sorry, I totally misread your comment! I like the idea of getting rid of just the time stamp 😄
